### PR TITLE
Add ANSI colors escapes to kdb

### PR DESCRIPTION
**Pull request by sirblackheart**
This is more of a suggestion. Not sure if you want colored output at all in kdb. There exist many different preferences in regard of colored terminal output.

For now the kdb info command is the only one where i found reason to add some highlights as it really can be a wall of text.
screenshot from 2016-05-16 18-08-02

### CODE REVIEW

**Comment by markus2330** [comment]
Actually I like it but there should be some configuration option to turn it off (e.g. if text is written to a file). Especially the error messages would really benefit from having some bold and colorful parts.

(in src/bindings/cpp/examples/cpp_example_userio.cpp you see how to use USER_DEFINED_IO, in src/bindings/cpp/include/kdbio.hpp is the original implementation currently used).

**Comment by sirblackheart** [comment]
> but there should be some configuration option to turn it off 

done

> (in src/bindings/cpp/examples/cpp_example_userio.cpp you see how to use USER_DEFINED_IO, in src/bindings/cpp/include/kdbio.hpp is the original implementation currently used).

so ansicolor should move into the libs so it can be used by kdb and the bindings?

**Comment by markus2330** [comment]
The PR looks fine except that the `doc/help/list` does not mention the option.

No, ansicolor should stay in `kdb`, but a new printWarnings/printError that (optionally) uses colors can be added to `kdb`.

**Comment by markus2330** [comment]
add to whitelist

**Comment by sirblackheart** [comment]
> No, ansicolor should stay in kdb, but a new printWarnings/printError that (optionally) uses colors can be added to kdb.
Working on it. So should colored output in general be enabled or disabled?

**Comment by sirblackheart** [comment]
Sorry pushed to fast, many test fail now because of the changed output.

**Comment by markus2330** [comment]
Dont worry about failing tests in the PRs, that is the reason why we have the build server.

Programs typically (e.g. ls) have an auto detection (default) and a way to turn it off/on.
With this convention it would be a string parameter with never/always/auto (or tty).
It would also be good to have a possibility to set a default via KDB.

`isatty (STDOUT_FILENO)` would be the code for detection.

Do not forget to run clang-format. Watch out that clang does not swap the header files (coloredkdbio) in a way that it does not work anymore. Use an empty file to enforce some order of header files (or define USER_DEFINED_IO outside).

**Comment by sirblackheart** [comment]
Not sure if the windows code works or not, i was not yet able to compile elektra on windows.

**Comment by markus2330** [comment]
@omnidan review please

**Comment by mpranj** [comment]
jenkins build mingw64 please

**Comment by sirblackheart** [comment]
@mpranj thank you

**Comment by mpranj** [comment]
np :) it appears that `STDERR_FILENO` and `STDOUT_FILENO` are being redefined on mingw64.
http://build.libelektra.org:8080/job/elektra-gcc-configure-mingw-w64/475/consoleFull

**Comment by sirblackheart** [comment]
So far the results with the windows color implementations are:
Colors including auto works for _isatty(_fileno (stdout)) but not for  _isatty(_fileno (stderr)) on the windows command line.
Colors work for stdout in mintty but not for stderr and stdout is not a tty in mintty.

**Comment by markus2330** [comment]
It seems like that there is a small formatting bug in src/tools/kdb/ansicolors.cpp

File: doc/AUTHORS [hunk 1]
```diff
@@ -38,6 +38,8 @@ Devel/Test on: Debian Stable
 Charles Lindsay
 ni library
 
+Gabriel Rauter <rauter.gabriel@gmail.com>
+xdg-icon theme, ansi colors in terminal kdb
```
Comment (markus2330) [inline doc/AUTHORS]: Its also ok to write about what you want to do, or something more general.

**Comment by markus2330** [comment]
Thank you, this contribution really improves usability! I will make a new video with the colors enabled ;)

@omnidan How is the review progressing? I would like to merge the PR.

**Comment by omnidan** [comment]
@markus2330 looks good to me :+1:

**Comment by markus2330** [comment]
Thank you!

**Pull request merged**